### PR TITLE
Add ADR settings API

### PR DESCRIPTION
**Pull request by adriansmares**
#### Summary

References https://github.com/TheThingsNetwork/lorawan-stack/issues/3820
References https://github.com/TheThingsNetwork/lorawan-stack/issues/5092

#### Changes

- Refactor the ADR settings API that is part of the end device MAC settings

#### Testing

This is purely API, N/A.

##### Regressions

N/A.

#### Notes for Reviewers

My proposal is to add the API mentioned in this PR, with the following semantics:
| |`use_adr` = `true` | `use_adr` = `false` | `use_adr`= `null`|
| --- | --- | --- | --- |
| `adr.mode` = `null` | Adapt the data rate, with the margin `adr_margin` (current behavior) | Use the desired parameters values (current behavior) | Adapt the data rate, with the margin `adr_margin` (current behavior) |
| `adr.mode` = `dynamic` | Adapt the data rate, with the margin `adr.mode.dynamic.margin` | Same as left hand side | Same as left hand side |
| `adr.mode` = `static` | Use the `adr.mode.static` parameters | Same as left hand side | Same as left hand side |
| `adr.mode` = `disabled` | Do not adapt the data rates, just enable/disable channels | Same as left hand side | Same as left hand side |

The idea is to change nothing for existing end devices, while introducing settings which are far more flexible for newer configurations. Once the end device is configured using `adr.mode`, the `use_adr` / `adr_margin` fields are ignored completely.

Note that the band has veto powers and may disable ADR completely. The ISM2400 band does this. For such bands we want to basically disable `adr.mode` = `dynamic`, since the algorithm is not compatible.

@htdvisser main reviewer, as @johanstokking is off for now.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: api/end_device.proto [hunk 1]
```diff
@@ -167,6 +167,39 @@ message EndDeviceVersion {
...
+    // Transmission power index to use.
+    uint32 tx_power_index = 2 [(validate.rules).uint32.lte = 15];
+    // Number of retransmissions.
+    uint32 nb_trans = 3 [(validate.rules).uint32.lte = 15];
+  }
+
+  // Configuration options for dynamic ADR.
+  message DynamicMode {
+    // The ADR margin tells the network server how much margin it should add in ADR requests.
+    // A bigger margin is less efficient, but gives a better chance of successful reception.
```
Comment (htdvisser) [inline api/end_device.proto]: Maybe add that if this is not set explicitly, the network default is used.

File: api/end_device.proto [hunk 2]
```diff
@@ -167,6 +167,39 @@ message EndDeviceVersion {
...
+    google.protobuf.FloatValue margin = 1;
+    // Maximum number of retransmissions.
+    google.protobuf.UInt32Value max_nb_trans = 2 [(validate.rules).uint32 = {gte: 1, lte: 3}];
```
Comment (htdvisser) [inline api/end_device.proto]: 1. Shouldn't we have both a `min_nb_trans` and `max_nb_trans`?
2. Should we then also have `{min,max}_{data_rate_index,tx_power_index}`?
3. Maybe add that if this is not set explicitly, the network default is used.

**Review by htdvisser — COMMENTED** [review]
I don't know if we need to define new messages, and "ADR modes". We can also start by adding additional constraints (`{min,max}_{data_rate_index,tx_power_index,nb_trans}`). That way we can also cover `static` and `dynamic`. Setting the existing `use_adr` to `false` should be equivalent to the `disabled` mode, right?

**Comment by adriansmares** [comment]
> I don't know if we need to define new messages, and "ADR modes". We can also start by adding additional constraints (`{min,max}_{data_rate_index,tx_power_index,nb_trans}`). That way we can also cover `static` and `dynamic`. Setting the existing `use_adr` to `false` should be equivalent to the `disabled` mode, right?

There are some problems with this, which I've described in the PR body. First of all, `use_adr` to `false` is not actually disabling ADR. It is currently impossible to disable ADR. Ever since https://github.com/TheThingsNetwork/lorawan-stack/pull/3604 , `use_adr` = `false` means that some external application needs to change the desired parameters to the current parameters in order to avoid having everything fall back to 0.

So `use_adr` setting currently has some very bad semantics, and there are customers in the wild that use them. At the same time, we cannot disable ADR at all using these settings. 

I don't think that we should add more boolean options in order to reach the state space that we want. Instead, we should prepare the API for future changes already - if you want to use static/custom values, just use the `static` field. The NS then doesn't have to look at historical traffic, build maps or anything like that - it will just attempt to send the MAC command to the end device until `current` = `desired`. The same argument comes about experimental / newer ADR implementations (ADR that lowers the data rate). Adding more values then doing a nested switch is worse in my opinion than making the `oneof` option clear.

For dynamic, we can also add more `min/max` values, but again these settings make sense only when you are actually using ADR - having them stored while ADR is disabled/static just makes validation harder. I can already reserve their indices, and we can implement them later on.

**Review by htdvisser — COMMENTED** [review]
Okay then I'm good with this.

File: api/end_device.proto [hunk 3]
```diff
@@ -167,6 +167,59 @@ message EndDeviceVersion {
...
+    // Number of retransmissions.
+    uint32 nb_trans = 3 [(validate.rules).uint32.lte = 15];
+  }
+
+  // Configuration options for dynamic ADR.
+  message DynamicMode {
+    // The ADR margin tells the network server how much margin it should add in ADR requests.
+    // A bigger margin is less efficient, but gives a better chance of successful reception.
+    // If unset, the default value from Network Server configuration will be used.
+    google.protobuf.FloatValue margin = 1;
```
Comment (johanstokking) [inline api/end_device.proto:187]: This is dB, right?

File: api/end_device.proto [hunk 4]
```diff
@@ -167,6 +167,59 @@ message EndDeviceVersion {
...
+    // Maximum transmission power index.
+    // If unset, the default value from Network Server configuration will be used.
+    google.protobuf.UInt32Value max_tx_power_index = 5 [(validate.rules).uint32.lte = 15];
+
+    // Minimum number of retransmissions.
+    // If unset, the default value from Network Server configuration will be used.
+    google.protobuf.UInt32Value min_nb_trans = 6 [(validate.rules).uint32 = {gte: 1, lte: 3}];
+    // Maximum number of retransmissions.
+    // If unset, the default value from Network Server configuration will be used.
+    google.protobuf.UInt32Value max_nb_trans = 7 [(validate.rules).uint32 = {gte: 1, lte: 3}];
```
Comment (johanstokking) [inline api/end_device.proto:208]: `NbTrans` has range `[1:15]`

**Review by johanstokking — APPROVED** [review]
API generally looks good, two remarks though that I'd like to see changed but no need for an additional review round.

Will there be a migration to move away from `use_adr`?

Also references https://github.com/TheThingsNetwork/lorawan-stack/issues/40 in the sense that those profiles could be a preset for the ADR mode. We may need additional (dynamic) options, like a soft minimum/maximum number of gateways, but we can do that later.

**Comment by adriansmares** [comment]
> Will there be a migration to move away from `use_adr`?

No, as there are probably API users in the wild. Instead, we will make the fields mutually exclusive (so using the API here requires that you don't use `use_adr`/`adr_margin`).

File: api/end_device.proto (hunk 4 above)
Comment (adriansmares) [reply api/end_device.proto:208]: The algorithm that we use does not go over `3`, so allowing values over 3 is going to have no effect, while being confusing. The static settings allow the constant value to go to 15.

**Pull request merged**